### ISSUE TITLE
Some strange behaviour with VH ready bit for DHV11

### ISSUE THREAD

**Issue by EJMarr** [simh/simh#609]
With a (historic) bespoke DHV11 driver, I am seeing the DHV11 RX.DATA.AVAIL bit set when I the driver does not expect it to be.

- #### Context

Because there was originally an issue with some serial line cards failing to interrupt, the driver regularly inspects the RX.DATA.AVAIL bit, and reports if it is set when not expected. It then tries to recover the situation by flipping RXIE, which should generate an interrupt if there is any data in the FIFO. It does not seem to do this properly. See detail below, and excruciating detail in the debug log and CPU history.

- #### the output of "sim> SHOW VERSION" while running the simulator which is having the issue
```
sim> sho version
PDP-11 simulator V4.0-0 Current
    Simulator Framework Capabilities:
        32b data
        32b addresses
        Threaded Ethernet Packet transports:PCAP:NAT:UDP
        Idle/Throttling support is available
        Virtual Hard Disk (VHD) support
        RAW disk and CD/DVD ROM support
        Asynchronous I/O support (Lock free asynchronous event queue)
        Asynchronous Clock support
        FrontPanel API Version 12
    Host Platform:
        Compiler: Microsoft Visual C++ 19.00.24215.01
        Simulator Compiled as C arch: x86 (Release Build) on Aug 12 2018 at 08:00:09
        Memory Access: Little Endian
        Memory Pointer Size: 32 bits
        Large File (>2GB) support
        SDL Video support: SDL Version 2.0.5
        PCRE RegEx (Version 8.36 2014-09-26) support for EXPECT commands
        OS clock resolution: 1ms
        Time taken by msleep(1): 1ms
        OS: Microsoft Windows [Version 10.0.17134.165]
        Architecture: x86 on AMD64, Processors: 2
        Processor Id: AMD64 Family 16 Model 4 Stepping 3, AuthenticAMD, Level: 16, Revision: 0403
        git commit id: 7eda7293
        git commit time: 2018-08-11T16:25:14-07:00
```
- #### how you built the simulator or that you're using prebuilt binaries

Clean build in VS2015, from VS2015 cloned source. No local modifications.

- #### the simulator configuration file (or commands) which were used when the problem occurred.

[Target2.ini.txt](https://github.com/simh/simh/files/2288318/Target2.ini.txt)

- #### the expected behavior and the actual behaviour

[T2VH.TXT](https://github.com/simh/simh/files/2288321/T2VH.TXT)

The DHV11 is configured for modem control, to be connected to from Reflections.

It is intended to carry out various internal initialisations on connection, but for the moment the only application code reports on the modem bits, and nothing else.

The application starts when the system starts, reporting once per second. The first time, it reports that the received modem bits are set, thereafter reporting them as clear (as one would expect).

When the Telnet connection from Reflections is made, the driver begins reporting that the DHV is failing to interrupt, as it sees the RX.DATA.AVAIL bit set.

Going into detail:

The driver carries out its reads at PC 142512; you can see several successful checks (before connection) at 20:05:10.336 (RX_DATA_AVAIL0 is shown).

At 20:05:18.192, the modem connection is made, the modem signals are shown, I'm guessing some Telnet information is passed, and the "Connected to" message is sent.

A receive interrupt is generated; it looks like some error bits are set, and RXIE is cleared and set by the driver.

At 20:05:18.298, 24 bytes which look like Telnet control are received. No interrupt is generated, and the message "Line:0 Remaining 0 bytes ''" suggests they have been handled internally.

Nothing else is logged until 20:05:18.865, when the driver finds RX.DATA.AVAIL1. It does not look to me as if it ought to be set at this point. RXIE is toggled, which should generate an interrupt if there is any data in the FIFO, but it does not attempt to do so. Note that interrupts are enable at this point - see CPU history at the end.

In the absence of any terminal input, this behaviour on the regular check continues indefinitely, not stopping on disconnection.

In this example, I attempted some input to see if there was any effect (pressing A several times).

This generates an interrupt, and it can be seen being read from the data buffer on the four occasions it was typed. However, it did not stop the problem.

The driver in this example has been modified to halt on the third occurrence of finding RX.DATA.AVAIL set, and the ini file then dumps the instruction history, so at 20:05:20.809 the simulation deliberately halts. Most of the history is various other parts of the system executing, but it is possible to tie up the log and the history for the various register accesses using the PC, and examine the PSW at the time.

Is management of RX.DATA.AVAIL correct?

**Comment by EJMarr** [comment]
I've solved the RX.DATA.AVAIL problem.

In fifo_get in pdp11_vh.c the bit is reset at the beginning, if the fifo is empty. This is not what the manual says. There is no mention of having to do a read, but it says the bit is cleared when the fifo becomes empty.

I changed the check at the beginning to remove the clear of the bit

```
	if (rbuf_idx[vh] == 0) { /* just be a bit defensive */
/*		vh_csr[vh] &= ~CSR_RX_DATA_AVAIL; */
		return (0);
	}
```

and added the if statement before the following original lines shown below.

```
	/* clear the CSR bit immediately */
	if (rbuf_idx[vh] == 0) {
		vh_csr[vh] &= ~CSR_RX_DATA_AVAIL;
	}
    /* Reschedule the next poll preceisely so that the 
       programmed input speed is observed. */
    sim_clock_coschedule_abs (vh_poll_unit, tmxr_poll);
    return (data & 0177777);
}
```

The driver I am using does not keep reading the fifo until it gets an empty character, but instead toggles RXIE to generate the next interrupt. This does not generate an interrupt in this case, as the fifo is empty, but regardless, the RX.DATA.AVAIL bit remains wrongly set.

I have also tested this change on RSX11M-PLUS and it still works.

This change does not affect the lack of generation of a modem state change interrupt on modem disconnect.

**Comment by markpizz** [comment]
That change seems reasonable.  It has been committed to the master branch.

I'm traveling until late on August 21, and won't have time to look closely at the other details you mentioned until I get back home.

There is support for direct connections to individual lines on the MUX.  See HELP VH ATTACH.

Meanwhile, I'd appreciate it if you could pull down the latest master branch code with:
```
> git reset --hard HEAD^
> git pull
```
Open the simh.sln file and let the conversion complete and then build the PDP11 simulator.  That should automatically realize that your windows-build is outdated and then AUTOMATICALLY pick up the right stuff.  Please let me know if you see anything unexpected.

**Comment by EJMarr** [comment]
I’m getting errors

 

1>------ Build started: Project: PDP11, Configuration: Release Win32 ------

1>  Library support for Visual Studio 2015 is available

1>     Creating library ..\BIN\NT\Win32-Release\PDP11.lib and object ..\BIN\NT\Win32-Release\PDP11.exp

1>SDL2-Static.lib(SDL_assert.obj) : error LNK2001: unresolved external symbol ___iob_func

1>SDL2-Static.lib(SDL_log.obj) : error LNK2001: unresolved external symbol ___iob_func

1>..\BIN\NT\Win32-Release\PDP11.exe : fatal error LNK1120: 1 unresolved externals

========== Build: 0 succeeded, 1 failed, 1 up-to-date, 0 skipped ==========

**Comment by EJMarr** [comment]
Moved over to VS2008 for the time being.

Having tracked down where VH sets the modem bits for a telnet connection, and failed to track down anywhere it attempts to unset the bits, I inserted some code doing roughly comparable inverse actions, but it needs to check all lines, I think. I'm not sure that this code is applicable to serial lines, which would need more thought.

In vh_svc(), towards the end, after

```
    /* interrupt driven in a real DHQ */
    tmxr_poll_rx (&vh_desc);
```

I added

```
	/* The above poll may detect and partially handle one or more modem 
	   disconnections but we need to do VH specific actions. We need to set
	   lp->lstat, and queue a modem change character for any modem enabled line. 
	   Code below only valid for telnet connections. Not sure about serial lines.*/
    for (line = 0; line < vh_desc.lines; line++){
		vh = line/VH_LINES;
        lp = &vh_parm[line];
		/* if connectable, but not connected, check if the DH line modem bits are set */
		if (((!lp->tmln->modem_control) || (lp->tmln->modembits & TMXR_MDM_DTR)) && !lp->tmln->conn ) {
			/* This is a modem enabled line */
			if (lp->lstat && (STAT_DSR | STAT_DCD | STAT_CTS | STAT_RI )){
				/* set up the line status register */
			    lp->lstat &= ~(STAT_DSR | STAT_DCD | STAT_CTS | STAT_RI );
                /* and queue the information in the fifo if required */
				if (lp->lnctrl & LNCTRL_LINK_TYPE)
                    fifo_put (vh, lp, RBUF_DIAG |
                        RBUF_PUTLINE (line) | ((lp->lstat >> 8) & 0376));

			}
		}
	}
```

after

```
    /* interrupt driven in a real DHQ */
    tmxr_poll_rx (&vh_desc);
```

This works for me (as proof of concept), but having previously identified that the connection / disconnection  seems to work only once, giving a connection failure afterwards, it still does this with the change inserted.

The message from Reflections is in a little window saying
```
Host:localhost
An error occurred in communications
Communications or network error occurred.
```

I don't get any errors with RSX. I'll carry on chasing it.

**Comment by EJMarr** [comment]
Forget the connection problem - it was a setting in the Reflections configuration for the bespoke system. I modified it while scratching around trying to figure out how to connect to a specific line, and that made it work once only.

**Comment by markpizz** [comment]
The disconnect signaling should be OK now as well as the windows build for all compilers.

For complete modem signalling behavior you should:
```
set vh MODEM
set vh HANGUP
```

**Comment by EJMarr** [comment]
There is no indication at the DHV that the modem signals change at the disconnection event. Provided DTR is set by the system the connection works, and DCD is seen as set by the DHV. On disconnection there is no opposite change.

See attached debug extract.

[VH Close.txt](https://github.com/simh/simh/files/2321791/VH.Close.txt)

**Comment by markpizz** [comment]
You are correct.  There was a fundamental bug in tmxr_set_get_modembits() which never reported DCD    as 0 after a disconnect.  It should be working correctly now.

**Comment by EJMarr** [comment]
That seems to work for me, giving DCD as I connect and clearing it as  disconnect. I'm presuming CTS that you mentioned in the fix applies only to real serial lines.

One point on the connection topic - one of the features of my bespoke system is the ability to save a system's memory to disc, and afterwards it performs a 'hot' restart of the system. Both the save and the restart include bus reset, so should reset the DHV, dropping DTR. The application waits 3 seconds before setting the lines it uses into modem control mode, and then immediately sets DTR on them. The telnet sessions are not disconnected. I'll attempt to gather some debug information on this, but it may take me a few days to get round to doing so.

A second issue is that after the restart after the save, the DHV output becomes incredibly slow. I had a brief look at this earlier (while there were still modem signal issues) and felt that is could be due to simh timer issues, as interrupts seemed to be not generated following output. Again, I'll gather debug information, but this sounds too far away from the original issue that I'll start a new issue. There's always a possibility that it's a problem with my system's driver.

**Comment by markpizz** [comment]
> That seems to work for me, giving DCD as I connect and clearing it as disconnect. I'm presuming CTS that you mentioned in the fix applies only to real serial lines.

CTS tracks DCD when we're simulating modem signals using telnet sessions.  If the line is attach to a real serial port, then the incoming signals (DSR, DCD, CTS, RI) are sensed from the serial port.

As for dropping DTR.  DTR dropping only disconnects a telnet session if the VH device is set HANGUP.

As for your hot restart working,  I would presume that normally, the computer would actually turned off between when the save to disc happened and when a 'hot' restart happened later.  :-)  Meanwhile, I see that a bus reset should cause the VH device to master reset, which should result in DTR dropping.  I've now added that.  As for odd speed behavior, the output of SHOW MUX will help you see what TMXR believes the line settings are vs what you think they should be.

**Comment by EJMarr** [comment]
I'll try out the above fix. Thanks.

>As for your hot restart working, I would presume that normally, the computer would actually turned off between when the save to disc happened and when a 'hot' restart happened later. :-) 

After the save to disc (or it could be over Ethernet), it just restarts from the existing memory. I suspect it is now milliseconds (in the real world a second or two), but have not actually checked. It could later be restarted in exactly the same state using BOOT.

**Comment by EJMarr** [comment]
Yes, I now get the DTR drop. Thanks.

I also have more information on the speed issue.

Let me give more background:

The system is memory resident, and can be constructed off-line entirely, or may have a basic system constructed off-line and be completed on-line. The completed system may then save its memory as a bootable system, either over Ethernet or to local media. There are thus three methods of start.
1. Cold start: boot an off-line constructed system into a machine
1. Hot start A: following a save of memory, continue with the existing memory image 
1. Hot start B: boot a saved memory image into a machine
When used with simh modem connected lines, there are two options for the boot methods; the terminal emulators may be connected before or after boot.

I attach an annotated debug log, which is a cold start, followed by a save and hot start A, together with simh SHO commands at various places.

The software in the example system comprises
1. A kernel, with DHV11 driver.
1. A graphics server generating Regis graphics commands for VT340 on VH lines 0 and 1. It sends initial set-up information to set the terminals into Regis mode, then waits for application programs to send it hardware independent graphic display information over comms, for translation and output to the serial lines.
1. A line control application, managing modem control signals on the two VH serial lines, via requests to the driver. 
1. Two graphics output applications, one for each of the display terminals.

At the cold start, the terminal emulators are pre-connected (after the ini file has run, but before the manually entered BOO). The line control application is set to carry out line monitoring after a 3 second initial delay. 

Summary/guide to the debug file.

**Cold** **Start**

1. The ini file runs and enables debugging, but does not include a boot command.
1. SHO VH DETAIL and SHO MUX are as expected
1. Two terminal connections were made, to VH lines 0 and 1
1. SHO VH DETAIL and SHO MUX do not show the connection yet
1. The system is booted from RQ
1. The driver executes a master reset of the DHV11
1. The initial entry (before the 3 second delay) of the modem control application sends a message that is picked up by an EXPECT, to act as a marker for when applications can start. 
1. Diagnostic bytes are read from the DHV receive buffer

**Comment by EJMarr** [comment]
I hadn’t finished when some random keystroke submitted the form.

 

The completed comment will follow this evening. Mowing the grass in daylight had priority.

**Comment by markpizz** [comment]
Feel free to 'edit' the comment that got submitted early so everything you're reporting is in one place.

**Comment by EJMarr** [comment]
By the way, TBUFFAD2 is decoded with TBUFAD1 formatting, and shows (partly) as TBUFAD1.

In vh_rd()

```
    static BITFIELD* bitdefs[] = {vh_csr_bits, vh_rbuf_bits, vh_lpr_bits, vh_stat_bits,
                                  vh_lnctrl_bits, vh_tbuffad1_bits, vh_tbuffad1_bits, vh_tbuffct_bits};
```
should be (I suggest)
```
    static BITFIELD* bitdefs[] = {vh_csr_bits, vh_rbuf_bits, vh_lpr_bits, vh_stat_bits,
                                  vh_lnctrl_bits, vh_tbuffad1_bits, vh_tbuffad2_bits, vh_tbuffct_bits};
```

**Comment by markpizz** [comment]
Fixed the TBUFFAD2 decode.  Meanwhile, I'm waiting for you to finish your comments about the speed issue and provide the debug info you are commenting on.

**Comment by EJMarr** [comment]
I though I had ???


Edmund Marr
Mobile + 44 (0) 7712 680511

On 4 Sep 2018, at 18:46, Mark Pizzolato <notifications@github.com> wrote:

Fixed the TBUFFAD2 decode. Meanwhile, I'm waiting for you to finish your comments about the speed issue and provide the debug info you are commenting on.

**Comment by EJMarr** [comment]
I agree, they are not there. I’m not sure what happened.

 

The comments themselves are gone, as I typed them into the comments page. The debug file I can’t get to until Friday. Realistically Saturday for the commentary.

 

A brief summary of what I found. Not all are necessarily faults – they are just the way I found things work.

 

Background:

 

The system includes

 

1.       A real-time kernel

2.       A DHV11 driver

3.       A graphics server, driving lines 0 and 1 of the DHV11. It is not itself modem aware. The server starts automatically on start up , sending packets to the terminals it presumes are connected.

4.       A modem management application, currently incomplete. All it does at the moment is watch DCD and when it sees it, it sets DTR. In the longer term, it will tell the graphics server to resend its initialisation, and it will also restart …

5.       Two graphics applications, which send machine independent graphics commands to the graphics server, for translation and output to the two VH lines.

 

The system can be built up off-line. In the example I analysed, the modem management application was also set to start, while the graphics applications were not.

 

On a cold start, booting the constructed system image from RQ

 

I started simh with an ini file that did not contain a boot command, did some SHO commands and then connected my two terminal emulator sessions.

There was no change to SHO MUX or SHO VH output, but the terminal emulators showed themselves as connected, while simh said it was not.

I then did the BOOT of the system that had not run before.

 

1.       Until the incoming connections are established, the transmit command are processed but nothing is actually sent.

2.       The incoming connection for line 0 arrived after 30ms, but that for line 1 was around 1 second later

3.       I then manually started the two graphics applications, and the VT340 graphics commands were output with a small number of milliseconds per message, consistent with speed=9600*32. Lacking the initialisation string the actual displays were somewhat in error, but that is to be expected.

4.       After that, the system memory was saved back to the location of the original boot image, and the system resumed at its restart entry point.

 

The system begins with a RESET instruction (I believe the save has one as well, though I have not explicitly checked this century), and then cancels all the transactions it had active. It tells each driver to reset all its devices, and restarts all the system processes and applications, in their requested manner, either a full restart, or entry at a trap address.

 

This activity can be seen taking place.

 

What is different about the hot start is that 

1.       The terminal lines were configured with NOHANGUP, so the connections are still in place

2.       The initial outputs are actually sent, which they were not on cold start

3.       The graphics applications are active and are restarted from the beginning, so output a significant amount immediately, resulting in interleaved I/O, in particular, because the initialisation strings are identical for the two lines, they start at the same time (or as near as) but they seem to finish at the same time. Two interrupt initiations are seen on consecutive lines of the debug file, but only one actual interrupt occurs, identified in the CSR as line 0. Significantly later on, the CSR is read, and TXACTION is seen to be set, with line 1 identified. I looked in the DHV manual to clarify what should happen for simultaneous completions, but it seemed totally unhelpful.

4.       The driver appears to be carrying out the same sequences of operations, but the VH seems to have difficulty executing them. In general output time from initiation to interrupt is usually in the range 100-200ms, rather than the small numbers with the cold start, and they are not consistent with the number of bytes in the transmission.

5.       On one occasion, a 255 byte message was initiated, 86 bytes were sent, and the content was more consistent with the end of the message rather than the start, and the completion interrupt happened immediately following that.

 

I then broke into simh with ctrlE, did some more SHO commands, and terminated the debug log.

 

The SHO commands did not indicate any change in speed setting, but did have non-zero values for STALLED, with 5 on line 1 and upward of 30 on line 0.

 

To me, this looks like there is significant interference between the two lines of the VH, affecting timing and sequencing, rather than a speed setting problem. It might be triggered in the first place by simultaneous completion of transmission.

 

I’ll send the full details at the weekend, but the above may give a hint.

**Comment by EJMarr** [comment]
Updated information submitted by edit … see comment beginning "Yes, I now get the DTR drop. Thanks." from 8 days previous to this one.

Attachments are there.

**Comment by EJMarr** [comment]
Further test result.

Similar timing symptoms appear with a single line being used for the graphics.

**Comment by EJMarr** [comment]
Further (and cleaner) information on the output issue following SAVE is given in #646
